sumtool/ngram/summary_ngram_lookup.py

```diff
@@ -0,0 +1,409 @@
+                "case": LookupCase.match_found.value,
+                "match": list(matched_doc_idx.item()),
+            }
+
+    def lookup_summary_from_dataset(self, summary, n):
+        """
+        lookup given summary from the whole training dataset
+
+        Args:
+            summary: A String, summary generated from the document
+            n: An integer, the rank of the grams that are generated
+
+        Returns:
+            A List of dictionaries
+            Dictionary: {"ngram": Tuple, "case": int, "match": List}
+            - ngram: a tuple of word indices
+            - case: (one of the values of LookupCase), which category given query belongs to
+            - match: list of matched document indices, empty if no match
+        """
+        print("Looking up summary ngrams from the training set")
+        summary_indices = self.tokenizer.encode(summary, add_special_tokens=False)
+
+        ngram_df = self.ngrams_root[n].to_pandas()
+
+        result_dict_list = []
+        sum_ngrams = [summary_indices[k:] for k in range(n)]
+        for summary_ngram in zip(*sum_ngrams):
+            result_dict = self.lookup(summary_ngram, ngram_df)
+            result_dict_list.append(
+                {
+                    "ngram": summary_ngram,
+                    "case": result_dict["case"],
+                    "match": result_dict["match"],
+                }
+            )
+        return result_dict_list
+
+    def lookup_summary_from_document(self, summary, document, n):
+        """
+        lookup given summary from the given document
+
+        Args:
+            summary: A String, summary generated from the document
+            document: A String
+            n: An integer, the rank of the grams that are generated
+
+        Returns:
+            A List of dictionaries
+            Dictionary: {"ngram": Tuple, "case": int, "match_count": int}
+            - ngram: a tuple of word indices
+            - case: (one of the values of LookupCase), which category given query belongs to
+            - match_count: number of matched ngrams in the document, 0 if no match
+        """
+        print("Looking up summary ngrams from the given document")
+        summary_indices = self.tokenizer.encode(summary, add_special_tokens=False)
+        document_indices = self.tokenizer.encode(document, add_special_tokens=False)
+
+        document_ngram_dict = defaultdict(set)
+        doc_ngrams = [
+            document_indices[k:] for k in range(n)
+        ]  # [document_indices, document_indices[1:]]
+
+        for document_ngram in zip(*doc_ngrams):
+            if document_ngram in document_ngram_dict:
+                document_ngram_dict[document_ngram] += 1
+            else:
+                document_ngram_dict[document_ngram] = 1
+
+        # print(document_ngram_dict)
+
+        result_dict_list = []
+        sum_ngrams = [summary_indices[k:] for k in range(n)]
+
+        for summary_ngram in zip(*sum_ngrams):
```

Comment:
I use an enum to make the error code more understandable, but not sure if it's the best way.